arwen/host/execution.go

```diff
@@ -11,70 +11,129 @@ func (host *vmHost) doRunSmartContractCreate(input *vmcommon.ContractCreateInput
-		}
+		vmOutput = host.onExitDirectCreateOrCall(err, vmOutput)
 	}()
 
 	address, err := blockchain.NewAddress(input.CallerAddr)
 	if err != nil {
 		output.SetReturnCode(vmcommon.ExecutionFailed)
-		return vmOutput
+		return
 	}
 
 	runtime.SetVMInput(&input.VMInput)
 	runtime.SetSCAddress(address)
+
 	output.AddTxValueToAccount(address, input.CallValue)
 	storage.SetAddress(runtime.GetSCAddress())
 
-	err = metering.DeductInitialGasForDirectDeployment(input)
+	codeDeployInput := arwen.CodeDeployInput{
+		ContractCode:         input.ContractCode,
+		ContractCodeMetadata: input.ContractCodeMetadata,
+		ContractAddress:      address,
+	}
+
+	vmOutput, err = host.performCodeDeploy(codeDeployInput)
+	return
+}
+
+func (host *vmHost) onExitDirectCreateOrCall(err error, vmOutput *vmcommon.VMOutput) *vmcommon.VMOutput {
+	host.Runtime().CleanInstance()
+	arwen.RemoveAllHostContexts()
+
+	return host.overrideVMOutputIfError(err, vmOutput)
+}
+
+func (host *vmHost) overrideVMOutputIfError(err error, vmOutput *vmcommon.VMOutput) *vmcommon.VMOutput {
+	if err == nil {
+		return vmOutput
+	}
+
+	output := host.Output()
+
+	var message string
+	if err == arwen.ErrSignalError {
+		message = output.ReturnMessage()
+	} else {
+		message = err.Error()
+	}
+
+	return output.CreateVMOutputInCaseOfError(output.ReturnCode(), message)
+}
+
+func (host *vmHost) performCodeDeploy(input arwen.CodeDeployInput) (*vmcommon.VMOutput, error) {
+	runtime := host.Runtime()
+	metering := host.Metering()
+	output := host.Output()
+
+	err := metering.DeductInitialGasForDirectDeployment(input)
 	if err != nil {
 		output.SetReturnCode(vmcommon.OutOfGas)
-		return vmOutput
+		return nil, err
 	}
 
 	vmInput := runtime.GetVMInput()
 	err = runtime.CreateWasmerInstance(input.ContractCode, vmInput.GasProvided)
 	if err != nil {
 		output.SetReturnCode(vmcommon.ContractInvalid)
-		return vmOutput
+		return nil, err
 	}
 
 	err = runtime.VerifyContractCode()
 	if err != nil {
 		output.SetReturnCode(vmcommon.ContractInvalid)
-		return vmOutput
+		return nil, err
 	}
 
 	idContext := arwen.AddHostContext(host)
 	runtime.SetInstanceContextID(idContext)
-	defer func() {
-		runtime.CleanInstance()
-		arwen.RemoveHostContext(idContext)
-	}()
 
 	err = host.callInitFunction()
 	if err != nil {
 		output.SetReturnCode(vmcommon.FunctionWrongSignature)
-		return vmOutput
+		return nil, err
 	}
 
-	output.DeployCode(address, input.ContractCode)
-	vmOutput = output.GetVMOutput()
+	output.DeployCode(input)
+	vmOutput := output.GetVMOutput()
+	return vmOutput, nil
+}
+
+func (host *vmHost) doRunSmartContractUpgrade(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput) {
+	host.ClearStateStack()
+	host.InitState()
 
-	return vmOutput
+	runtime := host.Runtime()
+	output := host.Output()
+	storage := host.Storage()
+
+	var err error
+	defer func() {
+		vmOutput = host.onExitDirectCreateOrCall(err, vmOutput)
+	}()
+
+	runtime.SetVMInput(&input.VMInput)
```

Comment:
Use `runtime.InitStateFromContractCallInput()` instead, which also initializes some extra fields of the `runtimeContext`. See `doRunSmartContractCall()` for an example of context initialization.

Reply:
Fixed.